util.c

```diff
@@ -1932,6 +1932,58 @@ int mdmon_running(const char *devnm)
 	return 0;
 }
 
+/* wait_for_mdmon_control_socket() - Waits for mdmon control socket
+ * to be created within specified time.
+ * @devnm: Device for which mdmon control socket should start.
+ *
+ * In foreground mode, when mdadm is trying to connect to control
+ * socket it is possible that the mdmon has not created it yet.
+ * Give some time to mdmon to create socket. Timeout set to 10 sec.
+ */
+int wait_for_mdmon_control_socket(char *devname)
+{
+	char path[100], name[20];
+	struct sockaddr_un addr;
+	int retry_count = 0;
+	int sfd, rv, len;
+	char *c;
+
+	if (is_subarray(devname)) {
+		devname++;
+		c = strchr(devname, '/');
+		if (!c)
+			return -1;
+		len = c - devname;
+		memcpy(name, devname, len);
+		name[len] = '\0';
+	} else {
+		len = strlen(devname);
+		memcpy(name, devname, len);
+		name[len] = '\0';
+	}
+	snprintf(path, sizeof(path), "%s/%s.sock", MDMON_DIR, name);
+	sfd = socket(PF_LOCAL, SOCK_STREAM, 0);
+	if (sfd < 0)
+		return -1;
+
+	addr.sun_family = PF_LOCAL;
+	strcpy(addr.sun_path, path);
+
+	for (retry_count = 0; retry_count < 10; retry_count++) {
+		rv = connect(sfd, (struct sockaddr*)&addr, sizeof(addr));
+		if (rv < 0) {
+			sleep_for(0, MSEC_TO_NSEC(200), true);
+			continue;
+		}
+		break;
+	}
+
+	if (rv < 0)
+		pr_err("Failed to connect to control socket.\n");
+	close(sfd);
+	return rv;
```

Comment:
Please consider using mdadm_status_t for errors instead of raw int as a return value. Outside this we are not interested in connect return status.  We just need information that it failed so I think that `MDADM_STATUS_ERROR` is fine